From: Patrick McHardy <kaber@trash.net>
Subject: net: fib_rules: rearrange struct fib_rule
Patch-mainline: 2.6.33
Git-commit: d285834001df372f81045bb41092e54943e93c84
References: fate#310840
Acked-by: Jiri Bohac <jbohac@suse.cz>

The ifname member is only used to resolve interface names and is not needed
during rule lookups. The target and ctarget members however are used during
rule lookups and are currently located in a second cacheline.

Move ifname further to the end to make sure both target and ctarget are
located in the same cacheline as other members used during rule lookups.

The layout on 64 bit changes from:

struct fib_rule {
...
    u32                        table;                /*    56     4 */
    u8                         action;               /*    60     1 */

    /* XXX 3 bytes hole, try to pack */

    /* --- cacheline 1 boundary (64 bytes) --- */
    u32                        target;               /*    64     4 */

    /* XXX 4 bytes hole, try to pack */

    struct fib_rule *          ctarget;              /*    72     8 */
    struct rcu_head            rcu;                  /*    80    16 */
    struct net *               fr_net;               /*    96     8 */
};

to:

struct fib_rule {
...
    u32                        table;                /*    40     4 */
    u8                         action;               /*    44     1 */

    /* XXX 3 bytes hole, try to pack */

    u32                        target;               /*    48     4 */

    /* XXX 4 bytes hole, try to pack */

    struct fib_rule *          ctarget;              /*    56     8 */
    /* --- cacheline 1 boundary (64 bytes) --- */
    char                       ifname[16];           /*    64    16 */
    struct rcu_head            rcu;                  /*    80    16 */
    struct net *               fr_net;               /*    96     8 */

};

Signed-off-by: Patrick McHardy <kaber@trash.net>
Signed-off-by: David S. Miller <davem@davemloft.net>
---
 include/net/fib_rules.h |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/include/net/fib_rules.h
+++ b/include/net/fib_rules.h
@@ -12,7 +12,6 @@ struct fib_rule
 	struct list_head	list;
 	atomic_t		refcnt;
 	int			ifindex;
-	char			ifname[IFNAMSIZ];
 	u32			mark;
 	u32			mark_mask;
 	u32			pref;
@@ -21,6 +20,7 @@ struct fib_rule
 	u8			action;
 	u32			target;
 	struct fib_rule *	ctarget;
+	char			ifname[IFNAMSIZ];
 	struct rcu_head		rcu;
 	struct net *		fr_net;
 };
